extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,315 @@
+  {
+    byte[] result = null;
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      final byte[] serialized = jedis.get(itemKey);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (IOException ex) {
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] serialized;
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      Jedis shard = jedis.getShard(itemKey);
+
+      if (expiration > 0) {
+        shard.psetex(itemKey, expiration, serialized);
+      }
+      else {
+        shard.set(itemKey, serialized);
+      }
+
+      final byte[] setKey = CacheImplUtils.computeNamespaceHash(prefix, key.namespace).getBytes();
+      shard = jedis.getShard(setKey);
+      shard.sadd(setKey, key.key);
+    }
+    catch (IOException | JedisConnectionException e) {
+      Throwables.propagate(e);
+    }
+    catch (JedisException ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error putting cache item");
+    }
+  }
+
+  @Override
+  public Map<NamedKey, byte[]> getBulk(Iterable<NamedKey> keys)
+  {
+    final Map<byte[], NamedKey> keyLookup = Maps.uniqueIndex(
+      keys,
+      new Function<NamedKey, byte[]>()
+      {
+        @Override
+        public byte[] apply(
+          @Nullable NamedKey input
+        )
+        {
+          return CacheImplUtils.computeKeyHash(prefix, input).getBytes();
+        }
+      }
+    );
+
+    // Sometimes broker passes empty keys list to getBulk()
+    if (keyLookup.size() == 0) {
+      return ImmutableMap.of();
+    }
+
+    final Map<NamedKey, byte[]> results = Maps.newHashMap();
+    final byte[][] keysArray = Iterables.toArray(keyLookup.keySet(), byte[].class);
+    final List<byte[]> valuesList = new ArrayList<>(keysArray.length);
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+      for (byte[] itemKey : keysArray) {
+        final byte[] data = jedis.get(itemKey);
```

Comment:
you may want to look into using pipelined jedis operations to combine multiple gets into one redis call.